src/reanimated2/animation/transformationMatrix/matrixUtils.tsx

```diff
@@ -0,0 +1,369 @@
+type FixedLengthArray<
+  T,
+  L extends number,
+  PassedObject = [T, ...Array<T>]
+> = PassedObject & {
+  readonly length: L;
+  [I: number]: T;
+};
+
+export type AffiniteMatrix = FixedLengthArray<FixedLengthArray<number, 4>, 4>;
+
+export type AffiniteMatrixFlat = FixedLengthArray<number, 16>;
+
+export type TransformMatrixDecomposition = Record<
+  | 'translationMatrix'
+  | 'scaleMatrix'
+  | 'rotationMatrix'
+  | 'skewMatrix'
+  | 'scaleMatrix',
+  AffiniteMatrix
+>;
+
+interface TansformMatrixDecompositionWithAngles
+  extends TransformMatrixDecomposition {
+  rx: number;
+  ry: number;
+  rz: number;
+}
+
+export function isAffiniteMatrixFlat(x: unknown): x is AffiniteMatrixFlat {
+  'worklet';
+  return (
+    Array.isArray(x) &&
+    x.length === 16 &&
+    x.every((element) => typeof element === 'number')
+  );
+}
+
+export function isAffiniteMatrix(x: unknown): x is AffiniteMatrix {
+  'worklet';
+  return (
+    Array.isArray(x) &&
+    x.length === 4 &&
+    x.every(
+      (row) =>
+        Array.isArray(row) &&
+        row.length === 4 &&
+        row.every((element) => typeof element === 'number')
+    )
+  );
+}
+
+export function flatten(matrix: AffiniteMatrix): AffiniteMatrixFlat {
+  'worklet';
+  return matrix.flat() as AffiniteMatrixFlat;
+}
+
+export function unflatten(matrix: AffiniteMatrixFlat): AffiniteMatrix {
+  'worklet';
+  const result = [];
+  const matrixCopy = [...matrix];
+  while (matrixCopy.length) {
+    result.push(matrixCopy.splice(0, 4));
+  }
+  return result as AffiniteMatrix;
+}
```

Comment:
This function only supports 4x4 matrices, can we simply hard-code the calculations instead of using `while` loop?

Reply:
Sure!